### PR TITLE
Fix memory leaks in `SnackBar` [prod-leak-fix]

### PR DESCRIPTION
**Pull request by ValentinVignal**
Part of https://github.com/flutter/flutter/issues/141198

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

**Comment by ValentinVignal** [comment]
cc @polina-c

**Review by polina-c — APPROVED** [review]

**Comment by polina-c** [comment]
We have a 'Google testing' test failed. This means one of two: 
1. Public test coverage is not good enough and should be extended to cover the failure
2. The internal usage of the feature is wrong and should be fixed

Putting the test details here hoping they will help you to choose between the two.
Let me know if you need code of `SnackBarContainer` or you can repro the failure without it.

Test code:
```
      testWidgets(
          'should remove snack bar when popping back from other scaffold, '
          'if the cross-scaffold min duration has expired.', (tester) async {
        await pumpScaffoldRoutes(tester);
        store.updateState(ApplicationState((b) => b
          ..view.messaging.snackBar = SnackBarState(
            'message',
            shownAt: now,
            persistAcrossScaffolds: persistAcrossScaffolds,
          ).toBuilder()));
        await tester.pumpAndSettle();

        // Verify snack bar shown on route #1.
        expect(find.byKey(Key('route1')), findsOneWidget);
        expectSnackBar('message');

        // Navigate to route #2.
        now = now.add(Duration(milliseconds: 500));
        await tester.tap(find.byKey(Key('navigate.button')));
        await tester.pumpAndSettle();
        expect(find.byKey(Key('route2')), findsOneWidget);

        // If persistent, expect the snack bar to transfer across.
        if (persistAcrossScaffolds) {
          expectSnackBar('message');
        } else {
          expectNoSnackBar();
        }

        // Pop back to route #1.
        now = now.add(defaultSnackBarDuration);
        await tester.pageBack();
        await tester.pumpAndSettle();
        expect(find.byKey(Key('route1')), findsOneWidget);

        // Expect the snack bar to not linger on route #1.
        expectNoSnackBar();
      });
```

Error:
```
	Null check operator used on a null value
		The relevant error-causing widget was:
		AnimatedBuilder
		AnimatedBuilder:third_party/dart/flutter/lib/src/material/snack_bar.dart:859:28
		When the exception was thrown, this was the stack:
		#0      _SnackBarState.build.<anonymous closure> (third_party/dart/flutter/lib/src/material/snack_bar.dart:864:43) ...
      

```

**Comment by ValentinVignal** [comment]
That is a weird bug, it looks like an animation is not set when running the `build`, but they are supposed to be since it is done in the `initState` and `didUpdateWidget` 🤔 I wasn't able to reproduce the failure, would you have access to `pumpScaffoldRoutes`, `store`, `SnackBarState`, `ApplicationState` ?

**Comment by polina-c** [comment]
It is challenging to share internal code in a way that does not leak internal sensitive information. 

Let's try other approach.
Can you please add some debug time (in asserts like [here](https://github.com/flutter/flutter/blob/cc9ac7d13cccd24d72faab566f9fb59663c3c3d1/packages/flutter/lib/src/material/material.dart#L737)) protection for usage after disposal?
To the animation itself and _SnackBarState.
So that we can see what caused the failure clearer.

Let's see how many tests will fail in Flutter. If not many and they are easy to fix, we may want to keep it. And it may help us to figure out the internal failure.

**Comment by ValentinVignal** [comment]
Sure, I added the `assert`s in [perf: Add debugDispose to snackbar and curved animation](https://github.com/flutter/flutter/pull/147212/commits/0a845a8c7530bb8faaccb8e0d38a98eaa83348d8). Let's see what comes out of it

File: packages/flutter/lib/src/animation/animations.dart [hunk 1]
```diff
@@ -435,8 +436,15 @@ class CurvedAnimation extends Animation<double> with AnimationWithParentMixin<do
     return reverseCurve == null || (_curveDirection ?? parent.status) != AnimationStatus.reverse;
   }
 
+  bool _debugDisposed = false;
```
Comment (polina-c) [inline packages/flutter/lib/src/animation/animations.dart:439]: I think we can use isDisposed defined on 425.

**Comment by polina-c** [comment]
It seems states are used after disposal. Can it be that `Null check operator used on a null value` that was raised in G3 is caused by usage after disposal?

If yes, can we make the classes more resilient to usage after disposal?

**Comment by ValentinVignal** [comment]
> perf: Add debugDispose to snackbar and curved animation

Sure, I guess I should also remove what was done in [perf: Add debugDispose to snackbar and curved animation](https://github.com/flutter/flutter/pull/147212/commits/0a845a8c7530bb8faaccb8e0d38a98eaa83348d8) (the  debug time protection in asserts for usage after disposal?) ?

**Comment by polina-c** [comment]
> > perf: Add debugDispose to snackbar and curved animation
> 
> Sure, I guess I should also remove what was done in [perf: Add debugDispose to snackbar and curved animation](https://github.com/flutter/flutter/pull/147212/commits/0a845a8c7530bb8faaccb8e0d38a98eaa83348d8) (the debug time protection in asserts for usage after disposal?) ?

For now you may want to replace debug asserts with debug prints in case of violation, so that we see if usage after disposal is causing failures in google testing, in case they are still here.

**Comment by polina-c** [comment]
Converted the PR to draft, as it is not intended for merge at the moment.

**Comment by ValentinVignal** [comment]
@polina-c [ci: Add debug prints to investigate failing ci](https://github.com/flutter/flutter/pull/147212/commits/1d0766e98dff3d2621f5ec0992a45b62c8d4cd30) adds some logs, do they show something in the google internal tests?

**Comment by polina-c** [comment]
Here is the error:

```
The following _TypeError was thrown building AnimatedBuilder(listenable: AnimationController#178cc(▶
		0.400; for SnackBar)➩Cubic(0.40, 0.00, 0.20, 1.00), dirty, state: _AnimatedState#23f3b):
		Null check operator used on a null value
		The relevant error-causing widget was:
		AnimatedBuilder
		AnimatedBuilder:third_party/dart/flutter/lib/src/material/snack_bar.dart:874:28
		When the exception was thrown, this was the stack:
		#0      _SnackBarState.build.<anonymous closure> (third_party/dart/flutter/lib/src/material/snack_bar.dart:879:43)
		#1      ListenableBuilder.build (third_party/dart/flutter/lib/src/widgets/transitions.dart:1127:48)
		#2      _AnimatedState.build (third_party/dart/flutter/lib/src/widgets/transitions.dart:138:48)
		#3      StatefulElement.build (third_party/dart/flutter/lib/src/widgets/framework.dart:5593:27)
		#4      ComponentElement.performRebuild (third_party/dart/flutter/lib/src/widgets/framework.dart:5481:15)
		#5      StatefulElement.performRebuild (third_party/dart/flutter/lib/src/widgets/framework.dart:5644:11)
		#6      Element.rebuild (third_party/dart/flutter/lib/src/widgets/framework.dart:5197:7)
		#7      BuildOwner.buildScope (third_party/dart/flutter/lib/src/widgets/framework.dart:2905:19)
		#8      AutomatedTestWidgetsFlutterBinding.drawFrame (third_party/dart/flutter_test/lib/src/binding.dart:1418:19)
		#9      RendererBinding._handlePersistentFrameCallback (third_party/dart/flutter/lib/src/rendering/binding.dart:468:5)
		#10     SchedulerBinding._invokeFrameCallback (third_party/dart/flutter/lib/src/scheduler/binding.dart:1392:15)
		#11     SchedulerBinding.handleDrawFrame (third_party/dart/flutter/lib/src/scheduler/binding.dart:1313:9)
		#12     AutomatedTestWidgetsFlutterBinding.pump.<anonymous closure> (third_party/dart/flutter_test/lib/src/binding.dart:1273:9)
		#15     TestAsyncUtils.guard (third_party/dart/flutter_test/lib/src/test_async_utils.dart:71:41)
		#16     AutomatedTestWidgetsFlutterBinding.pump (third_party/dart/flutter_test/lib/src/binding.dart:1260:27)
```

**Comment by ValentinVignal** [comment]
🤔 I'm still confused with it haha.
I believe the snack bar is getting disposed and for some reason the `AnimatedBuilder` using `_heightAnimation` rebuilds.

I can try to add more prints.

Alternatively, I could do a check like that:

```dart
     if (_heightAnimation != null) {
      snackBarTransition = AnimatedBuilder(
        animation: _heightAnimation!,
        builder: (BuildContext context, Widget? child) {
          return Align(
            alignment: AlignmentDirectional.topStart,
            heightFactor: _heightAnimation!.value,
            child: child,
          );
        },
        child: snackBar,
      );
    } else {
      snackBarTransition = child;
    }
```

but that looks wrong to me

**Comment by polina-c** [comment]
I do not know what is right thing to do here. I am open for experiments in different directions directions. Feel free to start tackling them in parallel by creating number of draft PRs that are intended to solve the same problem, to check how tests will react. :)

**Comment by ValentinVignal** [comment]
Sure I can do that. Can you help me check if there are some useful logs from the commit [debug: Add more prints](https://github.com/flutter/flutter/pull/147212/commits/e9763ce23edf9a7f1c166b5393db21fb0162b59d) ?

**Comment by polina-c** [comment]
Checking.
For debug prints it would be helpful to prefix all them with something outstanding like `!!!!`, to make search easier.

**Comment by polina-c** [comment]
```
Snackbar.dispose called. _SnackBarState#e1f97
		Snackbar.dispose called. _SnackBarState#e5677
		00:01 [32m+1[0m: SnackBarContainer with persistAcrossScaffolds=true should provide a SnackBar action when configured[0m
		Snackbar.dispose called. _SnackBarState#cfaa6
		00:01 [32m+2[0m: SnackBarContainer with persistAcrossScaffolds=true should hide SnackBar action when allowSnackbarAction=false[0m
		Snackbar.dispose called. _SnackBarState#06f1c
20:30:31.000	U	00:01 [32m+3[0m: SnackBarContainer with persistAcrossScaffolds=true should only transfer persistent snack bar across scaffolds[0m
		Snackbar.dispose called. _SnackBarState#352da
		Snackbar.dispose called. _SnackBarState#938f5
		CurvedAnimation.value is used after being disposed. AnimationController#8c4d9(▶ 0.400; for SnackBar)➩Interval(0.72⋯1.0)➩Cubic(0.40, 0.00, 0.20, 1.00)ₒₙ/Threshold
		SnackBar._heightAnimation in AnimatedBuilder. _SnackBarState#938f5(lifecycle state: defunct, not mounted)
		SnackBar is disposed in AnimatedBuilder. _SnackBarState#938f5(lifecycle state: defunct, not mounted)
		SnackBar._heightAnimation in AnimatedBuilder. _SnackBarState#938f5(lifecycle state: defunct, not mounted)
		SnackBar is disposed in AnimatedBuilder. _SnackBarState#938f5(lifecycle state: defunct, not mounted)
		══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
		The following _TypeError was thrown building AnimatedBuilder(listenable: AnimationController#8c4d9(▶
		0.400; for SnackBar)➩Cubic(0.40, 0.00, 0.20, 1.00), dirty, state: _AnimatedState#fa9e2):
		Null check operator used on a null value
		The relevant error-causing widget was:
		AnimatedBuilder
		AnimatedBuilder:third_party/dart/flutter/lib/src/material/snack_bar.dart:875:28
		When the exception was thrown, this was the stack:
		#0      _SnackBarState.build.<anonymous closure> (third_party/dart/flutter/lib/src/material/snack_bar.dart:889:43)
		#1      ListenableBuilder.build (third_party/dart/flutter/lib/src/widgets/transitions.dart:1127:48)
		#2      _AnimatedState.build (third_party/dart/flutter/lib/src/widgets/transitions.dart:138:48)
		#3      StatefulElement.build (third_party/dart/flutter/lib/src/widgets/framework.dart:5593:27)
		#4      ComponentElement.performRebuild (third_party/dart/flutter/lib/src/widgets/framework.dart:5481:15)
```

**Comment by ValentinVignal** [comment]
It looks like [fix: Use ValueListenableBuilder](https://github.com/flutter/flutter/pull/147212/commits/c79ce630dac64b96eb47df3a0d2d726622928570) made the trick 😃

**Comment by ValentinVignal** [comment]
@polina-c Is there something else you want me to do on this?

**Pull request merged**